File: two pointers/happy_number.java
public class happy_number {
    
}
class Solution {
    public boolean isHappy(int n) {
        int fast=n, slow=n;
        while(true){
            slow = sumOfSquare(slow);
            fast = sumOfSquare(sumOfSquare(fast));
            if(slow==fast){
                break;
            }
        }
        if(slow==1){
            return true;
        }
        else{
            return false;
        }
    }
    public int sumOfSquare(int n){
        int output = 0;
        while(n>0){
            int digit = n%10;
            digit = (int)Math.pow(digit,2);
            output = output+digit;
            n = (int)Math.floor(n/10);
        }
        return output;
    }
}

File: two pointers/happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        visit = set()
        while n not in visit: # iterate until cycle detect or n becomes 1(return)
            visit.add(n)
            n = self.sumOfSquares(n)
            if n == 1:
                return True
        return False

    def sumOfSquares(self, n):
        output = 0
        while n:
            digit = n%10
            digit = digit**2
            output += digit
            n = n//10
        return output

class Solution2:
    def isHappy(self, n: int) -> bool:
        slow, fast = n, n
        while True:
            slow = self.sumOfSquares(slow)
            fast = self.sumOfSquares(self.sumOfSquares(fast))
            if slow == fast:
                break
        return slow==1

    def sumOfSquares(self, n):
        output = 0
        while n:
            digit = n%10
            digit = digit**2
            output += digit
            n = n//10
        return output